src/world/WeatherManager.php

```diff
@@ -0,0 +1,196 @@
+			$this->duration = $duration;
+			return;
+		}
+
+		$event = new WeatherChangeEvent($this->world, $this->current, $type);
+		$event->call();
+		if($event->isCancelled()){
+			return;
+		}
+
+		$this->current = $type;
+		$this->duration = $duration;
+		$this->tick = 0;
+
+		$this->applyWeatherToWorld();
+		$this->sendWeatherPackets();
+	}
+
+	public function tick() : void{
+		$this->tick++;
+
+		if($this->tick >= $this->duration && $this->duration > 0){
+			$this->autoChange();
+		}
+
+		// Random lightning during thunder
+		if($this->current === WeatherType::THUNDER && mt_rand(0, 200) === 0){
+			$players = $this->world->getPlayers();
+			if(count($players) === 0){
+				return;
+			}
+
+			$player = $players[array_rand($players)];
+			$spawn = $player->getPosition();
+			$x = (int) $spawn->x + mt_rand(-40, 40);
+			$z = (int) $spawn->z + mt_rand(-40, 40);
+			$y = $this->world->getHighestBlockAt($x, $z) ?? max(64, (int) $spawn->y);
+
+			$this->spawnLightning(new Vector3($x, $y, $z));
+		}
+
+		foreach($this->lightnings as $i => $lightning){
+			if(!$lightning->tick()){
+				unset($this->lightnings[$i]);
+			}
+		}
+	}
+
+	private function autoChange() : void{
+		$roll = mt_rand(1, 100);
+		if($roll <= 70){
+			$this->setWeather(WeatherType::CLEAR, mt_rand(6000, 12000));
+		}elseif($roll <= 90){
+			$this->setWeather(WeatherType::RAIN, mt_rand(4000, 9000));
+		}else{
+			$this->setWeather(WeatherType::THUNDER, mt_rand(2000, 6000));
+		}
+	}
+
+	private function applyWeatherToWorld() : void{
+		if($this->current === WeatherType::RAIN){
+			$this->world->setRainLevel(1.0);
+			$this->world->setThunderLevel(0.0);
+		}elseif($this->current === WeatherType::THUNDER){
+			$this->world->setRainLevel(1.0);
+			$this->world->setThunderLevel(1.0);
+		}else{
+			$this->world->setRainLevel(0.0);
+			$this->world->setThunderLevel(0.0);
+		}
+	}
+
+	private function sendWeatherPackets() : void{
+		foreach($this->world->getPlayers() as $player){
+			$pos = $player->getPosition();
+
+			switch($this->current){
+				case WeatherType::CLEAR:
+					$rainEvent = LevelEventPacket::create(LevelEvent::STOP_RAIN, 0, $pos);
+					$thunderEvent = LevelEventPacket::create(LevelEvent::STOP_THUNDER, 0, $pos);
+					break;
+
+				case WeatherType::RAIN:
+					$rainEvent = LevelEventPacket::create(LevelEvent::START_RAIN, 80000, $pos);
+					$thunderEvent = LevelEventPacket::create(LevelEvent::STOP_THUNDER, 0, $pos);
+					break;
+
+				case WeatherType::THUNDER:
+					$rainEvent = LevelEventPacket::create(LevelEvent::START_RAIN, 160000, $pos);
+					$thunderEvent = LevelEventPacket::create(LevelEvent::START_THUNDER, 200000, $pos);
+					break;
+
+				default:
+					$rainEvent = LevelEventPacket::create(LevelEvent::STOP_RAIN, 0, $pos);
+					$thunderEvent = LevelEventPacket::create(LevelEvent::STOP_THUNDER, 0, $pos);
+					break;
+			}
+
+			$player->getNetworkSession()->sendDataPacket($rainEvent);
+			$player->getNetworkSession()->sendDataPacket($thunderEvent);
```

Comment:
This can use `NetworkBroadcastUtils::broadcastPackets()` instead of sending the packets one by one